### PR TITLE
Oc 12286:Concurrent Users Within Same Browser Session

### PR DESCRIPTION
**Pull request by Taoli2018**

### CODE REVIEW

File: web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java [hunk 1]
```diff
@@ -126,6 +119,24 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
...
+            if(currentSessionUser !=null && !(currentSessionUser.equals(eb))) {
+            	request.getSession().invalidate();
+            	
+            }            
+        	// Place the last username attempted into HttpSession for views
+        	session = request.getSession(false);                    
+
+            if (session != null || getAllowSessionCreation()) {
+                request.getSession().setAttribute(SPRING_SECURITY_LAST_USERNAME_KEY, TextEscapeUtils.escapeEntities(username));
+            }
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java]: please revert this change and implement the fix where userB would be prevented from logging in if userA already has an active session

Comment (Taoli2018) [reply web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java]: done

File: web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java [hunk 2]
```diff
@@ -108,6 +110,28 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
...
+        
+        UserAccountBean currentSessionUser = null;
+        currentSessionUser = (UserAccountBean) session.getAttribute(SecureController.USER_BEAN_NAME);
+        
+        if(currentSessionUser !=null && !(currentSessionUser.getName().isEmpty()) && !(currentSessionUser.getName().equals(username))) {
+        	//setFilterProcessesUrl("/pages/login/login?action=errorSessionLocked");
+        	try {
+				response.sendRedirect(request.getContextPath() +"/pages/login/login?action=errorSessionLocked");
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java]: instead of printing stack trace like this, use a logger.

File: web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java [hunk 3]
```diff
@@ -108,6 +110,28 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
...
+				e.printStackTrace();
+			}
+        	return null;
+        	//throw new LockedException("Already have  other user logined in in the same computer, please wait");
+        }
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java]: can you clean up the commented code?

File: web/src/main/resources/org/akaza/openclinica/i18n/notes.properties [hunk 4]
```diff
@@ -18,7 +18,7 @@ X_event_wiht_definition = The study event with definition '
 a_new_user_account_has_been_created_for_you = A new user account has been created for you on the OpenClinica system.  Your login information follows:\n\n
 
 account_locked = Your User Name and Password combination has been Locked. Please contact your OpenClinica Administrator to have them unlock your account.
-
+session_locked = Unable logging in. Another user is already logged in on this web browser. Please have them log out before attempting to log in.
```
Comment (svadla-oc) [inline web/src/main/resources/org/akaza/openclinica/i18n/notes.properties]: please update the error message according to the updated AC

File: web/src/main/webapp/WEB-INF/jsp/login/login.jsp [hunk 5]
```diff
@@ -1,8 +1,9 @@
 <%@page contentType="text/html;charset=UTF-8" language="java" %>
 <%@ include file="/WEB-INF/jsp/taglibs.jsp" %>
+<%String action = request.getParameter("action");%>
 <!-- For Mantis Issue 6099 -->
 <jsp:useBean scope='session' id='userBean' class='org.akaza.openclinica.bean.login.UserAccountBean'/>
-    <c:if test="${userBean.name!=''}">
+    <c:if test="${userBean.name!='' && action!=null}">
```
Comment (svadla-oc) [inline web/src/main/webapp/WEB-INF/jsp/login/login.jsp]: it seems like the check should be "action == null". Please confirm

File: web/src/main/java/org/akaza/openclinica/control/submit/InitialDataEntryServlet.java [hunk 6]
```diff
@@ -43,6 +46,35 @@
...
+		  String submitted =(String) request.getParameter("submitted");
+		  String checkInputs = (String)request.getParameter("checkInputs");
+		  
+		  if(submitted !=null && submitted.equals("1") && checkInputs!=null && checkInputs.equals("1")) {
+			  for (int i = 0; i < userRoles.size(); i++) {
+		            StudyUserRoleBean studyRole = (StudyUserRoleBean) userRoles.get(i);
+
+					if(studyRole.getRole().equals(Role.MONITOR))
+					{
+						auth = false;
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/control/submit/InitialDataEntryServlet.java]: This will prevent users with Monitor access to any of the studies from entering data. We should check specific study's user role.

File: web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java (hunk 2 above)
Comment (Taoli2018) [reply web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java]: done

File: web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java (hunk 3 above)
Comment (Taoli2018) [reply web/src/main/java/org/akaza/openclinica/web/filter/OpenClinicaUsernamePasswordAuthenticationFilter.java]: done

File: web/src/main/resources/org/akaza/openclinica/i18n/notes.properties (hunk 4 above)
Comment (Taoli2018) [reply web/src/main/resources/org/akaza/openclinica/i18n/notes.properties]: done

File: web/src/main/webapp/WEB-INF/jsp/login/login.jsp (hunk 5 above)
Comment (Taoli2018) [reply web/src/main/webapp/WEB-INF/jsp/login/login.jsp]: done

File: web/src/main/java/org/akaza/openclinica/control/submit/InitialDataEntryServlet.java (hunk 6 above)
Comment (Taoli2018) [reply web/src/main/java/org/akaza/openclinica/control/submit/InitialDataEntryServlet.java]: done
